### PR TITLE
ARROW-4084: [C++] Make Status static method support variadic arguments

### PR DESCRIPTION
**Pull request by fsaintjacques**
- Static constructors like `Status::Invalid` now supports variadic
arguments à la `Status::Invalid("my", variable, "error message: ", i)`.

- A new macro was added `ARROW_RETURN_IF(cond, status)` which replaces
the previous `ARROW_RETURN_IF_FALSE` but also adds branch prediction
hints. Note that only gandiva was refactored with this macro as
otherwise the code review would have exploded.

- Fixed a bug in memory map implementations not checking the return
code of `mmap` and `mremap`.

### CODE REVIEW

File: cpp/src/arrow/util/string_builder.h [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+namespace arrow {
+namespace util {
+
+template <typename Head>
+void StringBuilderRecursive(std::stringstream& stream, Head head) {
+  stream << head;
+}
+
+template <typename Head, typename... Tail>
+void StringBuilderRecursive(std::stringstream& stream, Head head, Tail... tail) {
```
Comment (bkietz) [inline cpp/src/arrow/util/string_builder.h]: The copies will probably be optimized out in this case, but idiomatically this should use perfect capture

Comment (bkietz) [reply cpp/src/arrow/util/string_builder.h]:
```
template <typename Head, typename... Tail>
void StringBuilderRecursive(std::stringstream& stream, Head&& head, Tail&&... tail) {
  StringBuilderRecursive(stream, std::forward<Head>(head));
  StringBuilderRecursive(stream, std::forward<Tail>(tail)...);
}
```

Comment (wesm) [reply cpp/src/arrow/util/string_builder.h]: +1 for using perfect forwarding

**Review by wesm — COMMENTED** [review]
This is definitely a lot cleaner syntax. Do you know if this impacts code size in any meaningful way? I will build the libraries locally to see optimized library size before and after

**Comment by wesm** [comment]
Impact on code size seems < 1% in a release build, libarrow increased by about 90K (total library size a little over 11MB for me with clang-6). Adding `inline` to the templates in string_builder.h made the code size a tiny bit smaller.

File: cpp/src/arrow/util/string_builder.h (hunk 1 above)
Comment (fsaintjacques) [reply cpp/src/arrow/util/string_builder.h]: Time to remove dust on my Effective modern C++ and read it!

File: cpp/src/arrow/status.h [hunk 2]
```diff
@@ -149,84 +145,106 @@ class ARROW_EXPORT Status {
   static Status OK() { return Status(); }
 
   /// Return a success status with a specific message
-  static Status OK(const std::string& msg) { return Status(StatusCode::OK, msg); }
+  template <typename... Args>
+  static Status OK(Args... args) {
+    return Status(StatusCode::OK, util::StringBuilder(args...));
```
Comment (pitrou) [inline cpp/src/arrow/status.h]: Do we need to use forwarding for `args...` here (and in the factories)?

File: cpp/src/arrow/util/string_builder.h [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  StringBuilderRecursive(stream, head);
+  StringBuilderRecursive(stream, tail...);
+}
+
+template <typename... Args>
+std::string StringBuilder(Args... args) {
+  std::stringstream stream;
+
+  StringBuilderRecursive(stream, args...);
```
Comment (pitrou) [inline cpp/src/arrow/util/string_builder.h]: Need to use forwarding here too?

**Review by pitrou — COMMENTED** [review]
Looks like a worthwhile improvement in readability :-)

**Comment by fsaintjacques** [comment]
Error is conda failing to download.

**Comment by wesm** [comment]
It times out sometimes, just restarted

**Comment by wesm** [comment]
+1

**Pull request closed without merging**